File: apps/starry/cpu-opencv-test/programs/carpets/cpp/opencv_mat.cpp
/* opencv_mat - cv::Mat arithmetic & structure vs element-exact closed form.
 *
 * add / subtract / multiply (per-element) / matmul (gemm) / transpose on KNOWN small matrices, asserted
 * element-exact against a hand-computed golden; plus Mat type/channels/ROI/reshape semantics. No RNG.
 */
#include "cv_common.h"
#include <opencv2/imgproc.hpp>

int main() {
    cv::setNumThreads(1);
    Gate g("OPENCV_MAT");

    /* Known 2x3 / 3x2 integer matrices. */
    cv::Mat A = (cv::Mat_<double>(2, 3) << 1, 2, 3, 4, 5, 6);
    cv::Mat B = (cv::Mat_<double>(2, 3) << 6, 5, 4, 3, 2, 1);

    /* element-wise add: A+B == all 7s */
    cv::Mat S = A + B;
    bool add_ok = true;
    for (int i = 0; i < 2; i++) for (int j = 0; j < 3; j++)
        add_ok &= close_d(S.at<double>(i, j), 7.0, 1e-9);
    g.check(add_ok, "A+B != 7 everywhere");

    /* element-wise subtract: A-B == [-5,-3,-1;1,3,5] */
    cv::Mat D = A - B;
    double sub_gold[2][3] = {{-5, -3, -1}, {1, 3, 5}};
    bool sub_ok = true;
    for (int i = 0; i < 2; i++) for (int j = 0; j < 3; j++)
        sub_ok &= close_d(D.at<double>(i, j), sub_gold[i][j], 1e-9);
    g.check(sub_ok, "A-B mismatch");

    /* element-wise multiply (Hadamard): A.mul(B) == [6,10,12;12,10,6] */
    cv::Mat M = A.mul(B);
    double mul_gold[2][3] = {{6, 10, 12}, {12, 10, 6}};
    bool mul_ok = true;
    for (int i = 0; i < 2; i++) for (int j = 0; j < 3; j++)
        mul_ok &= close_d(M.at<double>(i, j), mul_gold[i][j], 1e-9);
    g.check(mul_ok, "A.mul(B) mismatch");

    /* scalar multiply: 2*A */
    cv::Mat A2 = 2.0 * A;
    bool s2_ok = true;
    for (int i = 0; i < 2; i++) for (int j = 0; j < 3; j++)
        s2_ok &= close_d(A2.at<double>(i, j), 2 * A.at<double>(i, j), 1e-9);
    g.check(s2_ok, "2*A mismatch");

    /* matmul (gemm): A(2x3) * A^T(3x2) = [[14,32],[32,77]] */
    cv::Mat G = A * A.t();
    double mm_gold[2][2] = {{14, 32}, {32, 77}};
    bool mm_ok = (G.rows == 2 && G.cols == 2);
    for (int i = 0; i < 2 && mm_ok; i++) for (int j = 0; j < 2; j++)
        mm_ok &= close_d(G.at<double>(i, j), mm_gold[i][j], 1e-9);
    g.check(mm_ok, "A*A^T (gemm) mismatch");

    /* transpose: A^T is 3x2 with A^T(j,i)==A(i,j) */
    cv::Mat T = A.t();
    bool t_ok = (T.rows == 3 && T.cols == 2);
    for (int i = 0; i < 2 && t_ok; i++) for (int j = 0; j < 3; j++)
        t_ok &= close_d(T.at<double>(j, i), A.at<double>(i, j), 1e-9);
    g.check(t_ok, "transpose mismatch");

    /* determinant of a known 2x2: det([[1,2],[3,4]]) = -2 */
    cv::Mat K = (cv::Mat_<double>(2, 2) << 1, 2, 3, 4);
    g.check(close_d(cv::determinant(K), -2.0, 1e-9), "det != -2");

    /* inverse: K * K^-1 == I */
    cv::Mat Ki = K.inv();
    cv::Mat I2 = K * Ki;
    bool inv_ok = close_d(I2.at<double>(0, 0), 1, 1e-9) && close_d(I2.at<double>(1, 1), 1, 1e-9) &&
                  close_d(I2.at<double>(0, 1), 0, 1e-9) && close_d(I2.at<double>(1, 0), 0, 1e-9);
    g.check(inv_ok, "K*K^-1 != I");

    /* type / channels: a 3-channel 8U image */
    cv::Mat C8 = cv::Mat::zeros(4, 5, CV_8UC3);
    g.check(C8.type() == CV_8UC3, "type != CV_8UC3");
    g.check(C8.channels() == 3, "channels != 3");
    g.check(C8.rows == 4 && C8.cols == 5, "shape != 4x5");
    g.check(C8.elemSize() == 3, "elemSize != 3 bytes");

    /* ROI is a view: writing a ROI mutates the parent at the mapped location and nowhere else. */
    cv::Mat P = cv::Mat::zeros(6, 6, CV_8UC1);
    cv::Mat roi = P(cv::Rect(1, 2, 3, 2));   /* x=1,y=2,w=3,h=2 */
    roi.setTo(200);
    bool roi_ok = true;
    for (int y = 0; y < 6; y++) for (int x = 0; x < 6; x++) {
        int expect = (x >= 1 && x < 4 && y >= 2 && y < 4) ? 200 : 0;
        roi_ok &= (P.at<uchar>(y, x) == expect);
    }
    g.check(roi_ok, "ROI write did not map back to parent exactly");

    /* reshape: 12 contiguous values reshaped 1x12 -> 3x4 keeps row-major order. */
    cv::Mat lin(1, 12, CV_32S);
    for (int i = 0; i < 12; i++) lin.at<int>(0, i) = i;
    cv::Mat r34 = lin.reshape(1, 3);
    bool rs_ok = (r34.rows == 3 && r34.cols == 4);
    for (int i = 0; i < 3 && rs_ok; i++) for (int j = 0; j < 4; j++)
        rs_ok &= (r34.at<int>(i, j) == i * 4 + j);
    g.check(rs_ok, "reshape row-major order wrong");

    /* countNonZero / min-max on a known matrix. */
    cv::Mat mm = (cv::Mat_<uchar>(2, 3) << 0, 5, 0, 9, 0, 3);
    g.check(cv::countNonZero(mm) == 3, "countNonZero != 3");
    double mn, mx; cv::minMaxLoc(mm, &mn, &mx);
    g.check(close_d(mn, 0, 1e-9) && close_d(mx, 9, 1e-9), "minMax != (0,9)");

    return g.finish();
}

File: apps/starry/cpu-opencv-test/programs/carpets/py/opencv_mat.py
#!/usr/bin/env python3
"""opencv_mat - cv2/numpy Mat arithmetic & structure vs element-exact closed form.

add/subtract/multiply/matmul/transpose on KNOWN small matrices, asserted element-exact vs a numpy golden;
Mat dtype/channels/ROI/reshape semantics. cv2 Mats ARE numpy arrays, so numpy is the closed form. No RNG.
"""
import cv2
import numpy as np
from cv_common import Gate

cv2.setNumThreads(1)
g = Gate("OPENCV_MAT")

A = np.array([[1., 2., 3.], [4., 5., 6.]])
B = np.array([[6., 5., 4.], [3., 2., 1.]])

# element-wise add via cv2.add == numpy A+B == all 7
S = cv2.add(A, B)
g.check(np.array_equal(S, A + B) and np.all(S == 7.0), "cv2.add != A+B / not all 7")

# element-wise subtract via cv2.subtract
D = cv2.subtract(A, B)
g.check(np.array_equal(D, np.array([[-5., -3., -1.], [1., 3., 5.]])), "cv2.subtract mismatch")

# element-wise multiply via cv2.multiply (Hadamard)
M = cv2.multiply(A, B)
g.check(np.array_equal(M, np.array([[6., 10., 12.], [12., 10., 6.]])), "cv2.multiply mismatch")

# scalar scale
g.check(np.array_equal(cv2.multiply(A, 2.0), 2.0 * A), "scalar multiply mismatch")

# matmul (cv2.gemm): A(2x3) * A^T(3x2) = [[14,32],[32,77]]
G = cv2.gemm(A, A.T, 1.0, None, 0.0)
g.check(np.array_equal(G, np.array([[14., 32.], [32., 77.]])), "cv2.gemm mismatch")
# and numpy matmul agrees
g.check(np.array_equal(G, A @ A.T), "gemm != numpy @")

# transpose
T = cv2.transpose(A)
g.check(T.shape == (3, 2) and np.array_equal(T, A.T), "transpose mismatch")

# determinant of known 2x2
K = np.array([[1., 2.], [3., 4.]])
g.check(abs(cv2.determinant(K) - (-2.0)) < 1e-9, "det != -2")

# inverse: K @ K^-1 == I
ok, Ki = cv2.invert(K)
g.check(ok != 0.0 and np.allclose(K @ Ki, np.eye(2), atol=1e-9), "K*K^-1 != I")

# dtype / channels: 3-channel 8U image
C8 = np.zeros((4, 5, 3), dtype=np.uint8)
g.check(C8.dtype == np.uint8, "dtype != uint8")
g.check(C8.shape == (4, 5, 3), "shape != (4,5,3)")
g.check(C8.ndim == 3 and C8.shape[2] == 3, "channels != 3")

# ROI is a view: mutate parent at mapped location only
P = np.zeros((6, 6), dtype=np.uint8)
P[2:4, 1:4] = 200  # y=2..4, x=1..4
expect = np.zeros((6, 6), dtype=np.uint8)
expect[2:4, 1:4] = 200
g.check(np.array_equal(P, expect), "ROI write mapping wrong")

# reshape row-major
lin = np.arange(12, dtype=np.int32).reshape(1, 12)
r34 = lin.reshape(3, 4)
g.check(r34.shape == (3, 4) and np.array_equal(r34, np.arange(12).reshape(3, 4)), "reshape order wrong")

# countNonZero / minMaxLoc
mm = np.array([[0, 5, 0], [9, 0, 3]], dtype=np.uint8)
g.check(cv2.countNonZero(mm) == 3, "countNonZero != 3")
mn, mx, _, _ = cv2.minMaxLoc(mm)
g.check(mn == 0.0 and mx == 9.0, "minMax != (0,9)")

raise SystemExit(g.finish())
